### PR TITLE
kie-issues#114: JavaCodeCompletion module fails to compile because of API change

### LINKED ISSUE #114: JavaCodeCompletion module fails to compile because of API change
It seems that an API change on JDT.LS project is incompatible with our codebase.

Screenshot 2023-05-11 at 12 49 09

To solve that, I'm going to split this into multiple sub-tasks:
- Using the tag version 1.23.0 instead of the snapshot one. That will solve the issue. 
- Unfortunately, tagged versions get stale after some months (see https://github.com/eclipse/eclipse.jdt.ls/issues/1970) So, we need to update our codebase accordingly with the API changes

```[tasklist]
### Acceptance criteria
- [ ] https://github.com/kiegroup/kie-tools/pull/1626
- [ ] https://github.com/kiegroup/kie-tools/pull/1663
- [ ] https://github.com/kiegroup/kie-tools/pull/1688
- [ ] Update our codebase accordingly with the API changes (Using TypeHierarchy message?)
```

### PR DESCRIPTION
**Pull request by yesamer**
Finally closes https://github.com/apache/incubator-kie-issues/issues/114

One year ago the JDT.LS API suddenly changed breaking the Java class "Fetching" mechanism we had in place.
We that change, it was no longer possible to retrieve a class FQCN, which is a required property for the "Fetching" mechanism to properly work. 
In the latest release, the API now supports the TypeHierarchy command: given a class name, that command can retrieve the class hierarchy and the first items represent the class itself in FQCN format. 
To restore the "Fetching" feature, we can then rely on this new command.

For example, if we use that command on a String type variable, this will return:
Screenshot 2024-06-12 at 10 46 23
The first item contains exactly what we need.

https://github.com/apache/incubator-kie-tools/assets/16005046/15f0c4bc-f772-4b9f-93af-6a427edf25c4

### CODE REVIEW

**Comment by yesamer** [comment]
DO NOT MERGE - I need to double-check check if it works with JDT.LS 1.30.x

**Comment by yesamer** [comment]
It works with version 1.30.1! @tiagobento Ready to be reviewed and eventually merged.

File: packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java [hunk 1]
```diff
@@ -77,9 +75,35 @@ public BuildInformation buildPublicContent(Path filePath, String fqcn, String co
...
+        int index = 0;
+
+        for (int i = 0; i < line.length(); i++) {
+            if (line.charAt(i) != ' ') {
+                index = i;
+                break;
+            }
+        }
+
+        return index + 1;
```
Comment (jomarko) [inline packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java]:
```suggestion
        int index = line.indexOf(line.trim());

        return index + 1;
```

Was doing some quick 'googling' for the task of searching the first non-empty character in string, and found this proposal, that I quite like for its shortness, what do you think?

File: packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java [hunk 2]
```diff
@@ -77,9 +75,35 @@ public BuildInformation buildPublicContent(Path filePath, String fqcn, String co
...
+    public BuildInformation buildVarTypePublicContent(Path filePath, String fqcn, String completeText) {
+
+        TemplateParameters item = new TemplateParameters();
+        item.setClassName(getClassName(filePath));
+        item.setQuery(completeText);
+        item.setFqcn(fqcn);
+
+        String content = this.evaluate(Templates.TEMPLATE_ACCESSORS, item);
+
+        return new BuildInformation(filePath, getContent(filePath), content, 5, getFirstCharInLinePosition(content, 5) + 2);
```
Comment (jomarko) [inline packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java]: what are these numbers `5` and `2`, maybe we could store them as constant with some name?

**Review by jomarko — COMMENTED** [review]
Screenshot 2024-06-18 082502
Screenshot 2024-06-18 082455
Thank you @yesamer . I can confirm the import java class works.

I put two questions inline, and one question here. on the first wizard page, is that expected items are reordered once some item is selected?

File: packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java (hunk 2 above)
Comment (yesamer) [reply packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java]: Done

File: packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java (hunk 1 above)
Comment (yesamer) [reply packages/vscode-java-code-completion-extension-plugin/vscode-java-code-completion-extension-plugin-core/src/main/java/org/kogito/core/internal/engine/JavaEngine.java]: Nice improvement, thank you @jomarko . Done

**Comment by yesamer** [comment]
@jomarko Yep, the ones you select are put at the top of the list

**Review by jomarko — APPROVED** [review]
Thank you @yesamer

**Comment by yesamer** [comment]
@tiagobento Can you please review it?

**Comment by tiagobento** [comment]
Hi @yesamer! Can you please address the conflicts?

**Comment by yesamer** [comment]
I have some difficulties rebasing this... opening a new PR

**Pull request closed without merging**